[Reviewer: Andy] Preserve old DNS records if the server doesn't respond

rkd-msw opened the pull request:
We discussed earlier how the existing code probably doesn't behave how it's meant to (because records expire on schedule, so by the time we've realised that the DNS server hasn't responded, we've deleted the old records). This should improve that.

The basic change is that we add 5 minutes to the expiry time when working out when to free records, so there's a period where it has expired (and we'll re-query the DNS server) but is still in memory (and so we can re-use it if the server is down).

I also spotted that the existing code sets `ce->expires = 30` but I think it should be an epoch timestamp - I added a check to `add_to_expiry_list` to ensure this.

I haven't tested this yet (or even compiled it), but wanted to get it out for review to check the approach. In particular, this is meant to preserve existing records indefinitely if the server stays down - do you think that's right?

I'll test by:
* checking that my new warning isn't triggered during normal operation
* checking that DNS records are re-queried and updated during normal operation
* checking that if a DNS server goes down, the DNS record is preserved in the cache
* checking that if the server comes back with a different DNS record, that's picked up
* checking that NXDOMAIN responses don't preserve the existing cached record
* checking that if a record doesn't exist at start-of-day, everything still works

bossmc commented on src/dnscachedresolver.cpp:
```diff
@@ -545,8 +562,18 @@ DnsCachedResolver::DnsCacheEntryPtr DnsCachedResolver::create_cache_entry(const
+  {
+    LOG_WARNING("Cache expiry time is %d - expecting either 0 or an epoch timestamp (> %d)",
+                ce->expires,
+                sensible_minimum);
```
Can we hit this due to bogus external inputs?  If not, it's a pure code bug and (I reckon) we shouldn't try to recover... it's better to explode and drop a core for diagnosis/debugging.

rkd-msw replied:
Sensible in principle, but that's not the approach we follow elsewhere (e.g. https://github.com/Metaswitch/sprout/blob/dev/sprout/stack.cpp#L358) and might go against our current trend of trying to crash less (because crashing wipes out all other transactions).

Worth polling the team, though - if everyone agrees that's the right approach, I agree that this is one of the places where we should do that (but we should check all the LOG_WARNING and LOG_ERRORs too).

bossmc commented:
That looks like the right approach, your test plan sounds sensible too.

rkd-msw commented:
- [ ] checking that my new warning isn't triggered during normal operation
- [ ] checking that DNS records are re-queried and updated during normal operation
- [ ] checking that if a DNS server goes down, the DNS record is preserved in the cache
- [ ] checking that if the server comes back with a different DNS record, that's picked up
- [ ] checking that NXDOMAIN responses don't preserve the existing cached record
- [ ] checking that if a record doesn't exist at start-of-day, everything still works

rkd-msw commented:
Tested and fixed - can you re-review? Changes since the initial version are:
* special handling of NXDOMAIN
* moving the code into a common function, because it was duplicated

bossmc commented:
I've added some comments to https://github.com/Metaswitch/cpp-common/commit/748cc83a4a0bb00b8ac7008d4d502279f78f9b41, but otherwise. looks good.

rkd-msw commented:
OK - I've addressed your comment about asymmetry in the way we discussed (the single-domain dns_query now calls into the multiple-domain dns_query, so I don't need a common_query_logic function to avoid duplication between the two).

I haven't addressed your comment about it being a safe order - that's existing code, and orthogonal to this enhancement (but potentially worth digging into later).

I've also made _cache_lock a recursive mutex - ares_query can call the callback synchronously on the same thread in some error cases (including the ones this change is designed to cope with), which causes non-recursive mutexes to deadlock.

I've tested by making a call, stopping dnsmasq on Sprout, waiting for the TTL (60 secs), making another call, and checking that it succeeds.
